### PR TITLE
Fix ORM vs migration files inconsistencies

### PR DESCRIPTION
**Pull request by ephraimbuddy**
There have been some inconsistences between ORM and migration files but it doesn't fail in tests. This is an attempt to fix the inconsistency and also have it fail in tests

### CODE REVIEW

File: airflow/migrations/versions/0041_3_0_0_rename_dataset_as_asset.py [hunk 1]
```diff
@@ -285,13 +285,6 @@ def upgrade():
             unique=False,
         )
 
-        batch_op.create_foreign_key(
-            constraint_name="toar_asset_fkey",
-            referent_table="asset",
-            local_cols=["asset_id"],
-            remote_cols=["id"],
-            ondelete="CASCADE",
-        )
```
Comment (ephraimbuddy) [inline airflow/migrations/versions/0041_3_0_0_rename_dataset_as_asset.py:294]: I moved this to a new file because, for some reason, SQLite could not create this FK in this file. I suspect it is because the asset table was renamed in this migration

File: airflow/migrations/versions/0032_3_0_0_drop_execution_date_unique.py [hunk 2]
```diff
@@ -44,14 +45,24 @@
 
 def upgrade():
     with op.batch_alter_table("dag_run", schema=None) as batch_op:
-        batch_op.alter_column("execution_date", new_column_name="logical_date", existing_type=sa.TIMESTAMP)
+        batch_op.alter_column(
+            "execution_date",
+            new_column_name="logical_date",
+            existing_type=UtcDateTime,
```
Comment (kaxil) [inline airflow/migrations/versions/0032_3_0_0_drop_execution_date_unique.py]: I think we should use `airflow.migrations.db_types.TIMESTAMP` instead

https://github.com/apache/airflow/blob/5b2a96ee72a4bd7b7fbd4925c2f8c8468f047a64/airflow/migrations/db_types.py#L68

File: airflow/migrations/versions/0041_3_0_0_rename_dataset_as_asset.py (hunk 1 above)
Comment (kaxil) [reply airflow/migrations/versions/0041_3_0_0_rename_dataset_as_asset.py:294]: Yeah last I had tried I ran into:

```
Performing upgrade to the metadata database sqlite:////root/airflow/sqlite/airflow.db
[2024-10-27T22:43:35.430+0000] {migration.py:215} INFO - Context impl SQLiteImpl.
[2024-10-27T22:43:35.431+0000] {migration.py:218} INFO - Will assume non-transactional DDL.
[2024-10-27T22:43:35.432+0000] {migration.py:215} INFO - Context impl SQLiteImpl.
[2024-10-27T22:43:35.433+0000] {migration.py:218} INFO - Will assume non-transactional DDL.
[2024-10-27T22:43:35.433+0000] {db.py:1171} INFO - Creating tables
INFO [alembic.runtime.migration] Context impl SQLiteImpl.
INFO [alembic.runtime.migration] Will assume non-transactional DDL.
INFO [alembic.runtime.migration] Running upgrade 22ed7efa9da2 -> 044f740568ec, Drop ab_user.id foreign key.
INFO [alembic.runtime.migration] Running upgrade 044f740568ec -> d0f1c55954fa, Remove SubDAGs: `is_subdag` & `root_dag_id` columns from DAG table.
INFO [alembic.runtime.migration] Running upgrade d0f1c55954fa -> 0bfc26bc256e, Rename DagModel schedule_interval to timetable_summary.
INFO [alembic.runtime.migration] Running upgrade 0bfc26bc256e -> a2c32e6c7729, Add triggered_by field to DagRun.
INFO [alembic.runtime.migration] Running upgrade a2c32e6c7729 -> 1cdc775ca98f, Drop `execution_date` unique constraint on DagRun.
INFO [alembic.runtime.migration] Running upgrade 1cdc775ca98f -> 522625f6d606, Add tables for backfill.
INFO [alembic.runtime.migration] Running upgrade 522625f6d606 -> 16cbcb1c8c36, Remove redundant index.
INFO [alembic.runtime.migration] Running upgrade 16cbcb1c8c36 -> 44eabb1904b4, Update dag_run_note.user_id and task_instance_note.user_id columns to String.
INFO [alembic.runtime.migration] Running upgrade 44eabb1904b4 -> 0d9e73a75ee4, Add name and group fields to DatasetModel.
INFO [alembic.runtime.migration] Running upgrade 0d9e73a75ee4 -> c3389cd7793f, Add backfill to dag run model.
INFO [alembic.runtime.migration] Running upgrade c3389cd7793f -> 5a5d66100783, Add AssetActive to track orphaning instead of a flag.
INFO [alembic.runtime.migration] Running upgrade 5a5d66100783 -> fb2d4922cd79, Tweak AssetAliasModel to match AssetModel after AIP-76.
INFO [alembic.runtime.migration] Running upgrade fb2d4922cd79 -> 3a8972ecb8f9, Add exception_reason and logical_date to BackfillDagRun.
INFO [alembic.runtime.migration] Running upgrade 3a8972ecb8f9 -> 05234396c6fc, Rename dataset as asset.
/usr/local/lib/python3.9/contextlib.py:126 SAWarning: WARNING: SQL-parsed foreign key constraint '('alias_id', 'asset_alias', 'id')' could not be located in PRAGMA foreign_keys for table asset_alias_asset
/usr/local/lib/python3.9/contextlib.py:126 SAWarning: WARNING: SQL-parsed foreign key constraint '('dataset_id', 'asset', 'id')' could not be located in PRAGMA foreign_keys for table asset_alias_asset
Traceback (most recent call last):
File "/usr/local/lib/python3.9/site-packages/alembic/operations/batch.py", line 690, in drop_constraint
const = self.named_constraints.pop(const.name)
KeyError: 'dataset_alias_dataset_alias_id_fkey'
```

Comment (kaxil) [reply airflow/migrations/versions/0041_3_0_0_rename_dataset_as_asset.py:294]: Warnings:

```
/usr/local/lib/python3.9/contextlib.py:126 SAWarning: WARNING: SQL-parsed foreign key constraint '('alias_id', 'asset_alias', 'id')' could not be located in PRAGMA foreign_keys for table asset_alias_asset
/usr/local/lib/python3.9/contextlib.py:126 SAWarning: WARNING: SQL-parsed foreign key constraint '('dataset_id', 'asset', 'id')' could not be located in PRAGMA foreign_keys for table asset_alias_asset
/usr/local/lib/python3.9/contextlib.py:126 SAWarning: WARNING: SQL-parsed foreign key constraint '('alias_id', 'asset_alias', 'id')' could not be located in PRAGMA foreign_keys for table asset_alias_asset_event
/usr/local/lib/python3.9/contextlib.py:126 SAWarning: WARNING: SQL-parsed foreign key constraint '('event_id', 'asset_event', 'id')' could not be located in PRAGMA foreign_keys for table asset_alias_asset_event
```

File: airflow/migrations/versions/0041_3_0_0_rename_dataset_as_asset.py [hunk 3]
```diff
@@ -565,7 +551,7 @@ def downgrade():
     with op.batch_alter_table("task_outlet_dataset_reference", schema=None) as batch_op:
         batch_op.alter_column("asset_id", new_column_name="dataset_id", type_=sa.Integer(), nullable=False)
 
-        batch_op.drop_constraint("toar_asset_fkey", type_="foreignkey")
+        # batch_op.drop_constraint("toar_asset_fkey", type_="foreignkey")
```
Comment (kaxil) [inline airflow/migrations/versions/0041_3_0_0_rename_dataset_as_asset.py]: ?

File: airflow/migrations/versions/0048_3_0_0_add_dag_versioning.py [hunk 4]
```diff
@@ -131,6 +131,7 @@ def upgrade():
 
 def downgrade():
     """Unapply add dag versioning."""
+    _delete_serdag_and_code()
```
Comment (kaxil) [inline airflow/migrations/versions/0048_3_0_0_add_dag_versioning.py]: Does this fix any inconsistency? If not might be better as a separate PR

File: airflow/migrations/versions/0050_3_0_0_remove_pickled_data_from_xcom_table.py [hunk 5]
```diff
@@ -101,9 +101,9 @@ def upgrade():
         op.execute(
             """
             ALTER TABLE xcom
-            ALTER COLUMN value TYPE JSONB
+            ALTER COLUMN value TYPE JSON
             USING CASE
-                WHEN value IS NOT NULL THEN CAST(CONVERT_FROM(value, 'UTF8') AS JSONB)
+                WHEN value IS NOT NULL THEN CAST(CONVERT_FROM(value, 'UTF8') AS JSON)
```
Comment (kaxil) [inline airflow/migrations/versions/0050_3_0_0_remove_pickled_data_from_xcom_table.py]: handled in https://github.com/apache/airflow/pull/44290

File: airflow/migrations/versions/0050_3_0_0_remove_pickled_data_from_xcom_table.py [hunk 6]
```diff
@@ -136,6 +136,7 @@ def upgrade():
         # Drop the old `value_old` column
         with op.batch_alter_table("xcom", schema=None) as batch_op:
             batch_op.drop_column("value_old")
+    op.drop_table("_xcom_archive")
```
Comment (kaxil) [inline airflow/migrations/versions/0050_3_0_0_remove_pickled_data_from_xcom_table.py]: I didn't drop it purposely to avoid loss of data, was planning to handle it as part of `db cleanup`

File: airflow/migrations/versions/0050_3_0_0_remove_pickled_data_from_xcom_table.py [hunk 7]
```diff
@@ -101,9 +101,9 @@ def upgrade():
         op.execute(
             """
             ALTER TABLE xcom
-            ALTER COLUMN value TYPE JSONB
+            ALTER COLUMN value TYPE JSON
```
Comment (kaxil) [inline airflow/migrations/versions/0050_3_0_0_remove_pickled_data_from_xcom_table.py]:
```suggestion
            ALTER COLUMN value TYPE JSONB
```

File: airflow/migrations/versions/0050_3_0_0_remove_pickled_data_from_xcom_table.py (hunk 5 above)
Comment (kaxil) [inline airflow/migrations/versions/0050_3_0_0_remove_pickled_data_from_xcom_table.py]:
```suggestion
                WHEN value IS NOT NULL THEN CAST(CONVERT_FROM(value, 'UTF8') AS JSONB)
```

File: airflow/migrations/versions/0050_3_0_0_remove_pickled_data_from_xcom_table.py (hunk 6 above)
Comment (kaxil) [reply airflow/migrations/versions/0050_3_0_0_remove_pickled_data_from_xcom_table.py]: added support here: https://github.com/apache/airflow/pull/44291

File: airflow/migrations/versions/0032_3_0_0_drop_execution_date_unique.py (hunk 2 above)
Comment (ephraimbuddy) [reply airflow/migrations/versions/0032_3_0_0_drop_execution_date_unique.py]: What we have in execution_date is UtcDatetime before AF 3. This was changed when execution_date was changed to logical_date. I don't think we should change to Timestamp. Also, all of our date columns currently use UtcDatetime

File: airflow/migrations/versions/0050_3_0_0_remove_pickled_data_from_xcom_table.py (hunk 6 above)
Comment (ephraimbuddy) [reply airflow/migrations/versions/0050_3_0_0_remove_pickled_data_from_xcom_table.py]: A little fix here: https://github.com/apache/airflow/pull/44221/commits/47128e97fcc6614168c023a3f15db4c38a23eaf5

File: airflow/migrations/versions/0032_3_0_0_drop_execution_date_unique.py (hunk 2 above)
Comment (kaxil) [reply airflow/migrations/versions/0032_3_0_0_drop_execution_date_unique.py]: Those are the same. Both use `sa.TIMESTAMP`

https://github.com/apache/airflow/blob/5b2a96ee72a4bd7b7fbd4925c2f8c8468f047a64/airflow/utils/sqlalchemy.py#L48-L62

https://github.com/apache/airflow/blob/5b2a96ee72a4bd7b7fbd4925c2f8c8468f047a64/airflow/migrations/db_types.py#L43-L68

The difference IMO lies in `UtcDateTime` is mainly used in the Application/ORM layer (to ensure all dates are returned in UTC) while `db_types.TIMESTAMP` should be used for alembic migration.

Comment (ephraimbuddy) [reply airflow/migrations/versions/0032_3_0_0_drop_execution_date_unique.py]: I had some tests failing when the migration use TIMESTAMP instead of UtcDatetime. The tests under trigger_dagrun operator. I'll change it to TIMESTAMP to reveal those tests. 
I believe the reason those tests fail is that in this PR, the test database is initialized with the migration files which picks up the TIMESTAMP

Comment (kaxil) [reply airflow/migrations/versions/0032_3_0_0_drop_execution_date_unique.py]: Maybe it is worth consolidating them if it is possible

Comment (ephraimbuddy) [reply airflow/migrations/versions/0032_3_0_0_drop_execution_date_unique.py]: I imported the Timestamp from db_types instead of sa.TIMESTAMP. Let's see how it goes

Comment (ephraimbuddy) [reply airflow/migrations/versions/0032_3_0_0_drop_execution_date_unique.py]: Went well!

File: airflow/models/__init__.py [hunk 8]
```diff
@@ -63,7 +63,6 @@ def import_all_models():
     import airflow.models.serialized_dag
     import airflow.models.taskinstancehistory
     import airflow.models.tasklog
-    import airflow.providers.fab.auth_manager.models
```
Comment (jedcunningham) [inline airflow/models/__init__.py:66]: Do we no longer need this because FAB manages its own migrations?

Comment (ephraimbuddy) [reply airflow/models/__init__.py:66]: Yes. I guess I will do this in a separate PR

Comment (jedcunningham) [reply airflow/models/__init__.py:66]: I'm good either way, just wanted to make sure I knew why :)

**Review by jedcunningham — APPROVED** [review]

**Comment by ephraimbuddy** [comment]
Old sqlalchemy versions don't support drop table `if exists`. Should we have the _xcom_archive table in the ORM? cc @kaxil  see: https://github.com/apache/airflow/actions/runs/12034647961/job/33552179419?pr=44221#step:7:57121

**Comment by kaxil** [comment]
> Old sqlalchemy versions don't support drop table `if exists`. Should we have the _xcom_archive table in the ORM? cc @kaxil see: https://github.com/apache/airflow/actions/runs/12034647961/job/33552179419?pr=44221#step:7:57121

oh, huh,  we can remove the `if_exists` then

**Comment by ephraimbuddy** [comment]
> > Old sqlalchemy versions don't support drop table `if exists`. Should we have the _xcom_archive table in the ORM? cc @kaxil see: https://github.com/apache/airflow/actions/runs/12034647961/job/33552179419?pr=44221#step:7:57121
> 
> oh, huh, we can remove the `if_exists` then. We can use almebic get tables to do it
> 
> ```python
> def get_current_table_names():
>     connection = op.get_bind()
> 
>     metadata = MetaData()
>     metadata.reflect(bind=connection)
> 
>     return metadata.tables.keys()
> ```
> 
> or
> 
> https://github.com/apache/airflow/blob/761cedd06103e10627ad62727eae0aac15387130/airflow/utils/db.py#L1008-L1029

I think that would not work well with offline sql scripts. I’ll verify

**Comment by ephraimbuddy** [comment]
Didn't think of using raw SQL. I just used it now. Let's see how it goes

**Review by kaxil — APPROVED** [review]
Just one comment https://github.com/apache/airflow/pull/44221/files#r1854515334

File: airflow/migrations/versions/0041_3_0_0_rename_dataset_as_asset.py (hunk 1 above)
Comment (ephraimbuddy) [reply airflow/migrations/versions/0041_3_0_0_rename_dataset_as_asset.py:294]: Only one remains now, and I have spent quality time trying to solve it but have yet to figure it out. Since this is sqlite only, should we merge and debug the issue in another PR?
Remaining one:
```
/usr/local/lib/python3.9/contextlib.py:126 SAWarning: WARNING: SQL-parsed foreign key constraint '('alias_id', 'asset_alias', 'id')' could not be located in PRAGMA foreign_keys for table asset_alias_asset
```

Comment (kaxil) [reply airflow/migrations/versions/0041_3_0_0_rename_dataset_as_asset.py:294]: Sure, that sounds good to me

File: tests_common/test_utils/db.py [hunk 9]
```diff
@@ -73,7 +73,12 @@ def initial_db_init():
     from airflow.www.extensions.init_appbuilder import init_appbuilder
     from airflow.www.extensions.init_auth_manager import get_auth_manager
 
+    from tests_common.test_utils.compat import AIRFLOW_V_3_0_PLUS
+
     db.resetdb()
+    if AIRFLOW_V_3_0_PLUS:
+        db.downgrade(to_revision="5f2621c13b39")
+        db.upgradedb(to_revision="head")
```
Comment (ashb) [inline tests_common/test_utils/db.py:81]: Do we need to do this in every single test run, or could we limit this to the test job specifically checking the db upgrade/downgrade flow?

Comment (ephraimbuddy) [reply tests_common/test_utils/db.py:81]: I can’t see how to limit it to only the test that runs migration. My thinking is that we can’t limit it due to running tests in parallel and that doing it specifically for the /utils/db tests could affect other tests running. 

Also, the code only runs when we have —with-db-init flag on tests, which i think happens when tests are initialized and not on every one of the tests

Comment (potiuk) [reply tests_common/test_utils/db.py:81]: Yeah. One time only per session (and only when DB is not initialized or when `--with-db-init`)  - so optimising it is not really needed.

Comment (potiuk) [reply tests_common/test_utils/db.py:81]: BTW. We should indeed optimize some of the fixtures I think - but I'd say we should measure the impact and optimise those that are really slowing things down: see https://github.com/apache/airflow/issues/43488 -> otherwise `premature optimisation is the root of all evil (and complexity in the code)`
